src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const versionOfState = [];
+  let newState = { ...state };
+
+  for (const i of actions) {
+    switch (i.type) {
+      case 'addProperties':
+        Object.assign(newState, i.extraData);
+        versionOfState.push({ ...newState });
+        break;
+
+      case 'removeProperties':
+        for (const j of i.keysToRemove) {
+          if (j in newState) {
+            delete newState[j];
+          }
+        }
+        versionOfState.push({ ...newState });
+        break;
+
+      case 'clear' :
+        newState = {};
+        versionOfState.push({ ...newState });
```

Comment:
It would be great to add a default case to avoid potential errors in a situation where the type does not match any of the cases.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#syntax